addons/website_sale/models/sale_order.py

```diff
@@ -122,10 +122,11 @@ def _website_product_id_change(self, order_id, product_id, qty=0):
                     pu = price
         else:
             pu = product.price
-            if order.pricelist_id and order.partner_id:
-                order_line = order._cart_find_product_line(product.id)
-                if order_line:
-                    pu = self.env['account.tax']._fix_tax_included_price_company(pu, product.taxes_id, order_line[0].tax_id, self.company_id)
+
+        if order.pricelist_id and order.partner_id:
+            order_line = order._cart_find_product_line(product.id)
+            if order_line:
+                pu = self.env['account.tax']._fix_tax_included_price_company(pu, product.taxes_id, order_line[0].tax_id, self.company_id)
```

Comment:
Isn't that weird to compute the `pu` after the discount was calculated with another `pu` (line 117) ?

It's like:
- pu is X
- discount is based on pu which is X
- pu is not X actually, it's Y

Can you confirm the code that introduced that discount stuff is still working, as it was introduced without a test?
https://github.com/odoo-dev/odoo/commit/180866fe1e3b8998cc47ee584abaaff76d4135bb

Reply:
> But discount was computed on a wrong pu, isn't it?

Indeed, in some cases, the `pu` is incorrect. However, it does not impact the associated flow because currently this block is mainly used to compute the discount
For instance the test `:TestWebsiteSaleProductPricelist.test_cart_update_with_fpos`, without any fix:
When updating the cart
https://github.com/odoo/odoo/blob/a1beb1a4b3d7cc9724f5a9adeb1183c0a89e175f/addons/website_sale/tests/test_website_sale_product_attribute_value_config.py#L190-L189
prices and discounts are recomputed:
https://github.com/odoo/odoo/blob/a1beb1a4b3d7cc9724f5a9adeb1183c0a89e175f/addons/website_sale/models/sale_order.py#L110-L117
Considering the test, we have `price == 55`, `pu == 110` and `discount == 50`. The discount is the only correct value (since tax is now 0%, `pu` should be $100). But fortunately, `pu` is recalculated correctly afterwards and set with its correct value
https://github.com/odoo/odoo/blob/a1beb1a4b3d7cc9724f5a9adeb1183c0a89e175f/addons/website_sale/models/sale_order.py#L258-L263
(`pu` becomes $100) so everything is back to normal

So, I would say that either we keep the current fix (use a temporary `pu` and `price` to compute the discount, and at the end compute the real `pu`) or we apply your suggestion
> Naively, I'd say that the if condition you moved (unindented) should remain there, and the exact same condition/code should also be applied in the if (so not only in the else as it was) before the discount is computed.

In such case the fix would looks like https://gist.github.com/adwid/7906c9a8907fe143269b66e98ca29608

Reply:
As long as you validated that the
```
pu = something
discount = something(pu)
pu = something_else
```

Is actually not wrong, you can either keep your diff or do as I suggested, everything is fine to me :+1: (maybe if you keep your initial diff, add a comment line like `pu might not be the correct one and will be recomputed later to fix taxes, but discount can be computed on its current value as taxes do not alter the discount` (or something shorter ;-)))